### PR TITLE
Reference implementation of CT descriptors

### PR DESCRIPTION
**Pull request by apoelstra**
I'm aware that there are already a basic form of CT descriptors in this library, and this does not remove them, it just adds a new module which should act as a reference implementation for https://github.com/ElementsProject/ELIPs/pull/1

WIP because it needs a new rust-elements release to get tagged hashes in `bitcoin_hashes`.

### CODE REVIEW

**Comment by apoelstra** [comment]
Oops, missed some files. Added and force-pushed. I squashed everything because the individual commits weren't split out right; let me know if I should split them up again.

File: src/confidential/mod.rs [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+
+/// A description of a blinding key
+#[derive(Clone, PartialEq, Eq, PartialOrd, Ord, Debug, Hash)]
+pub enum Key<Pk: MiniscriptKey> {
+    /// Blinding key is computed from the hash of a sorted list of public keys
+    ///
+    /// The first public key is included outside of the `Vec` to reduce allocations
+    /// and to enforce that the list of keys is not empty; but it is not special in
+    /// any way.
+    HashToSk(Pk, Vec<Pk>),
```
Comment (sanket1729) [inline src/confidential/mod.rs:37]: This makes creating the `Key` really annoying. Can we just have a  Vec<Pk> instead? To ensure that there is at least one element, we should wrap this in a struct and add constructors.

Comment (sanket1729) [reply src/confidential/mod.rs:37]: IMO, we should really make things convenient for users by adding some extra code and allocations here.

File: src/confidential/mod.rs [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+    /// Obtains the blinded address for this descriptor.
+    pub fn address<C: secp256k1_zkp::Signing>(
+        &self,
+        secp: &secp256k1_zkp::Secp256k1<C>,
+        params: &'static elements::AddressParams,
+    ) -> Result<elements::Address, Error> {
+        let spk = self.descriptor.script_pubkey();
+        self.descriptor
+            .blinded_address(self.key.to_public_key(secp, &spk), params)
+    }
```
Comment (sanket1729) [inline src/confidential/mod.rs:116]: We should add other descriptor APIs later.

File: src/confidential/slip77.rs [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+        // taken from libwally src/test/test_confidential_addr.py
+        let mbk = MasterBlindingKey::from_seed(&unhex("c76c4ac4f4e4a00d6b274d5c39c700bb4a7ddc04fbc6f78e85ca75007b5b495f74a9043eeb77bdd53aa6fc3a0e31462270316fa04b8c19114c8798706cd02ac8"));
+        assert_eq!(mbk.as_bytes(), &unhex("6c2de18eabeff3f7822bc724ad482bef0557f3e1c1e1c75b7a393a5ced4de616")[..]);
+
+        let secp = secp256k1_zkp::Secp256k1::new();
+        let spk = elements::Script::from_hex("76a914a579388225827d9f2fe9014add644487808c695d88ac").unwrap();
+        let mut addr = elements::Address::from_str("2dpWh6jbhAowNsQ5agtFzi7j6nKscj6UnEr").unwrap();
+        addr.blinding_pubkey = Some(mbk.blinding_key(&secp, &spk));
+        assert_eq!(addr.to_string(), "CTEkf75DFff5ReB7juTg2oehrj41aMj21kvvJaQdWsEAQohz1EDhu7Ayh6goxpz3GZRVKidTtaXaXYEJ");
+    }
```
Comment (sanket1729) [inline src/confidential/slip77.rs:111]: I did not know this code, but I tested this with elements 22.0 and it works as expected. Perhaps, you can also add this test?
```rust
    #[test]
    fn local_test_elements_22_0() {
        // Local test on elements 22.0
        let mbk = MasterBlindingKey::from_hex(
            "64269a8de756da06ebe35d26dccb4dd46bddcf858b54eeaae315490cfe6cacc0",
        )
        .unwrap();

        let addr = elements::Address::from_str(
            "el1qqg2pz79c0reryhr6hzxrzueju9m2asllwydrhexs6vj854cvwlen4tryh4thsdt2a26rte3fe87rf3my9t90wt78pcqrxv733",
        )
        .unwrap();

        let derived_blinding_key = mbk.blinding_private_key(&addr.script_pubkey());
        assert_eq!(
            derived_blinding_key,
            secp256k1_zkp::SecretKey::from_slice(&unhex(
                "791a1081ae2ad98a5ad603737c648247f19d3c26e2beb54617638172edb230e7"
            ))
            .unwrap()
        );
    }
```

**Review by sanket1729 — COMMENTED** [review]
Should we remove the `Blinded` descriptors now? Should this struct also be called `Descriptor`?

File: src/confidential/mod.rs (hunk 1 above)
Comment (apoelstra) [reply src/confidential/mod.rs:37]: lol, okay, agreed

**Comment by apoelstra** [comment]
@sanket1729 regarding "should we remove the `Blinded` descriptors now", yes, and re "should this be called `Descriptor`, I think yes, the goal is to use it as `confidential::Descirptor`. We can also re-export it under a new name.

**Comment by sanket1729** [comment]
@apoelstra, agreed on both things

**Comment by apoelstra** [comment]
Addressed all comments. Remaning todo:

* Update reference impl to use `hash_to_private` (and address comments on that PR)
* Generate test vectors from the reference impl
* Add a note forbidding uncompressed keys

**Comment by apoelstra** [comment]
So, where I'm at with this is that I need
* a way to encode internal descriptors without checksums, which requires
* https://github.com/rust-bitcoin/rust-miniscript/issues/477 and then
* a new rust-miniscript release, whose use requires
* an update of the rust-miniscript dependency here, which requires
* an update of the rust-bitcoin and rust-elements dependencies here, which require
* https://github.com/ElementsProject/rust-elements/pull/152

**Comment by sanket1729** [comment]
Thanks for the enumerated list. Left some feedback on elements repo.

**Comment by sanket1729** [comment]
@apoelstra, what is the status of this project/PR?

**Comment by apoelstra** [comment]
@sanket1729 I intend to finish this and merge it. I think I've got everything I need now? Will try to do it over the weekend or early next week.

**Comment by apoelstra** [comment]
Rebase only.

**Comment by apoelstra** [comment]
cc @sanket1729 I have finally finished this PR, with test vectors and everything.

No rush, but ready for review.

**Comment by LeoComandini** [comment]
It would be nice to have `elements_miniscript::confidential::Descriptor::parse_descriptor` implemented so we can use descriptors strings with (extended) private keys. Now with `from_str` we can only parse descriptors with public keys. It would allow to create test vectors with private keys.

**Comment by apoelstra** [comment]
On @LeoComandini out-of-band advice we may modify this scheme so that the "bare pubkey" variant p2c-commits to the scriptpubkey. Conveniently, even when the scriptpubkey is exposed alongside the resulting key, nobody can "undo" the p2c and learn the original public key. So we get very good on-chain privacy here, even when one participant in a multiparty setup is reusing keys.

Will update after some more thought. Writing here so I don't forget.

**Comment by sanket1729** [comment]
Can now be rebased

**Comment by apoelstra** [comment]
Nice. I need to add some xpub-based test vectors but this is otherwise ready for review.

**Comment by apoelstra** [comment]
cc @sanket1729 ready for review

**Review by sanket1729 — APPROVED** [review]
ACK 725f4502c9e00a805f9a50cae58f64907e303530. 

Checked against the spec. Any reason why don't have wildcards in `BlindedKey`?

**Comment by apoelstra** [comment]
> Checked against the spec. Any reason why don't have wildcards in `BlindedKey`?

There isn't a type called `BlindedKey`, but I guess you mean, "why not have wildcards in the blinding key". The reason is a bit dumb -- it's because we have a generic keytype for the main descriptor, but for the blinding key we have either (a) a generic type, or (b) a `DescriptorSecretKey`.

Generic key types don't, in general, have wildcards. Normally, wildcards are handled by people taking a generic descriptor, using `DescriptorPublicKey` with it, and then using some convenience methods around `Descriptor::translate_pk` to map every pubkey to a `DefiniteDescriptorKey` at a particular index. The descriptor itself doesn't understand anything about indices.

So, when the descriptor has an extra "blinding key" which needs to be updated in tandem with the the rest of the keys, our generic-key model breaks down. I think the right way to handle this would be to link `DescriptorPublicKey` and `DescriptorSecretKey` somehow, so that we could say that CT view descriptors had a blinding key which was `Pk::SecretKey` or something rather than a concrete key type. Then for the specific case of `DescriptorPublicKey`, we could support deriving public and secret keys at the same time.

Does this make sense? Basically the issue is that we're mixing a generic keytype and a concrete one, and our API does not handle this gracefully.

**Comment by sanket1729** [comment]
@apoelstra, I get the issues with `DescriptorSecretKey`. One of the main benefits of decsriptors is backing up and recovery. 

Perhaps, I am misunderstanding something here, if `ct(xpub/*,wsh(ms))`, would the backups not grow linearly instead of just a single copy?

**Comment by apoelstra** [comment]
@sanket1729 I don't understand what you're saying. The spec allows wildcards. The reference implementation just doesn't support it because I didn't want to do the rearchitecture work required for it.

**Review by delta1 — COMMENTED** [review]
2 minor comment suggestions

**Review by sanket1729 — APPROVED** [review]

**Comment by apoelstra** [comment]
Going to merge this; can address comment nits in a followup PR.

**Pull request merged**